scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -1030,20 +1030,17 @@ int match_layer(u32 layer_id, pid_t pid, const char *cgrp_path)
 {
 
 	struct task_struct *p;
-	struct layer *layer;
-	u32 nr_match_ors;
+	struct layer *layer = &layers[layer_id];
```

Comment:
This won't pass the verifier as there are no bounds checks on if `layer_id` is valid. That is why it is set after the bounds check is done with `nr_match_ors`.
```
22:38:36 [WARN] libbpf: prog 'layered_runnable': -- BEGIN PROG LOAD LOG --
Validating match_layer() func#1...
285: R1=scalar() R2=scalar() R3=mem_or_null(id=49,sz=1) R10=fp0
; int match_layer(u32 layer_id, pid_t pid, const char *cgrp_path) @ main.bpf.c:1029
285: (bf) r8 = r3                     ; R3=mem_or_null(id=49,sz=1) R8_w=mem_or_null(id=49,sz=1)
286: (bc) w7 = w1                     ; R1=scalar() R7_w=scalar(smin=0,smax=umax=0xffffffff,var_off=(0x0; 0xffffffff))
; struct layer *layer = &layers[layer_id]; @ main.bpf.c:1033
287: (bc) w1 = w7                     ; R1_w=scalar(id=50,smin=0,smax=umax=0xffffffff,var_off=(0x0; 0xffffffff)) R7_w=scalar(id=50,smin=0,smax=umax=0xffffffff,var_off=(0x0; 0xffffffff))
288: (27) r1 *= 1061192               ; R1_w=scalar(smin=0,smax=umax=0x103147ffefceb8,smax32=0x7ffffff8,umax32=0xfffffff8,var_off=(0x0; 0x1ffffffffffff8))
289: (18) r9 = 0xffffc90080402000     ; R9_w=map_value(map=bpf_bpf.bss,ks=4,vs=16979080)
291: (0f) r9 += r1                    ; R1_w=scalar(smin=0,smax=umax=0x103147ffefceb8,smax32=0x7ffffff8,umax32=0xfffffff8,var_off=(0x0; 0x1ffffffffffff8)) R9_w=map_value(map=bpf_bpf.bss,ks=4,vs=16979080,smin=0,smax=umax=0x103147ffefceb8,smax32=0x7ffffff8,umax32=0xfffffff8,var_off=(0x0; 0x1ffffffffffff8))
; u32 nr_match_ors = layer->nr_match_ors; @ main.bpf.c:1034
292: (bf) r1 = r9                     ; R1_w=map_value(map=bpf_bpf.bss,ks=4,vs=16979080,smin=0,smax=umax=0x103147ffefceb8,smax32=0x7ffffff8,umax32=0xfffffff8,var_off=(0x0; 0x1ffffffffffff8)) R9_w=map_value(map=bpf_bpf.bss,ks=4,vs=16979080,smin=0,smax=umax=0x103147ffefceb8,smax32=0x7ffffff8,umax32=0xfffffff8,var_off=(0x0; 0x1ffffffffffff8))
293: (07) r1 += 1060992               ; R1_w=map_value(map=bpf_bpf.bss,ks=4,vs=16979080,off=0x103080,smin=0,smax=umax=0x103147ffefceb8,smax32=0x7ffffff8,umax32=0xfffffff8,var_off=(0x0; 0x1ffffffffffff8))
294: (61) r1 = *(u32 *)(r1 +0)
R1 unbounded memory access, make sure to bounds check any such access
processed 1089 insns (limit 1000000) max_states_per_insn 2 total_states 69 peak_states 69 mark_read 8
-- END PROG LOAD LOG --
```